File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/database/DigitalCashDatabaseTest.java
package com.github.dedis.popstellar.repository.database;

import androidx.test.core.app.ApplicationProvider;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.github.dedis.popstellar.di.AppDatabaseModuleHelper;
import com.github.dedis.popstellar.model.objects.Lao;
import com.github.dedis.popstellar.model.objects.digitalcash.TransactionObject;
import com.github.dedis.popstellar.model.objects.security.KeyPair;
import com.github.dedis.popstellar.repository.DigitalCashRepositoryTest;
import com.github.dedis.popstellar.repository.database.digitalcash.*;
import com.github.dedis.popstellar.testutils.Base64DataUtils;

import org.junit.*;
import org.junit.runner.RunWith;

import java.security.GeneralSecurityException;
import java.util.ArrayList;
import java.util.List;

import io.reactivex.observers.TestObserver;

import static com.github.dedis.popstellar.testutils.Base64DataUtils.generatePublicKey;

@RunWith(AndroidJUnit4.class)
public class DigitalCashDatabaseTest {

  private static AppDatabase appDatabase;
  private static TransactionDao transactionDao;
  private static HashDao hashDao;

  private static final String LAO_ID = Lao.generateLaoId(generatePublicKey(), 1000, "LAO1");
  private static final String LAO_ID2 = Lao.generateLaoId(generatePublicKey(), 1500, "LAO2");
  private static final KeyPair USER1 = Base64DataUtils.generateKeyPair();
  private static TransactionObject TRANSACTION_OBJECT;
  private static TransactionEntity TRANSACTION_ENTITY;
  private static HashEntity HASH_ENTITY =
      new HashEntity(USER1.getPublicKey().computeHash(), LAO_ID, USER1.getPublicKey());
  private static List<HashEntity> HASH_ENTITIES = new ArrayList<>();

  static {
    HASH_ENTITIES.add(HASH_ENTITY);
  }

  @Before
  public void before() throws GeneralSecurityException {
    appDatabase =
        AppDatabaseModuleHelper.getAppDatabase(ApplicationProvider.getApplicationContext());
    transactionDao = appDatabase.transactionDao();
    hashDao = appDatabase.hashDao();
    TRANSACTION_OBJECT =
        DigitalCashRepositoryTest.getValidTransactionBuilder("random id", USER1).build();
    TRANSACTION_ENTITY = new TransactionEntity(LAO_ID, TRANSACTION_OBJECT);
  }

  @After
  public void close() {
    appDatabase.close();
  }

  @Test
  public void insertTransactionTest() {
    TestObserver<Void> testObserver = transactionDao.insert(TRANSACTION_ENTITY).test();

    testObserver.awaitTerminalEvent();
    testObserver.assertComplete();
  }

  @Test
  public void getTransactionsTest() {
    TestObserver<Void> testObserver = transactionDao.insert(TRANSACTION_ENTITY).test();

    testObserver.awaitTerminalEvent();
    testObserver.assertComplete();

    TestObserver<List<TransactionObject>> testObserver2 =
        transactionDao
            .getTransactionsByLaoId(LAO_ID)
            .test()
            .assertValue(
                transactionObjects ->
                    transactionObjects.size() == 1
                        && transactionObjects.get(0).equals(TRANSACTION_OBJECT));

    testObserver2.awaitTerminalEvent();
    testObserver2.assertComplete();
  }

  @Test
  public void deleteTransactionsByLaoTest() {
    TestObserver<Void> testObserver = transactionDao.insert(TRANSACTION_ENTITY).test();

    testObserver.awaitTerminalEvent();
    testObserver.assertComplete();

    TestObserver<Void> testObserver2 = transactionDao.deleteByLaoId(LAO_ID).test();

    testObserver2.awaitTerminalEvent();
    testObserver2.assertComplete();

    // Assert that the transaction is actually deleted
    TestObserver<List<TransactionObject>> testObserver3 =
        transactionDao.getTransactionsByLaoId(LAO_ID).test().assertValue(List::isEmpty);

    testObserver3.awaitTerminalEvent();
    testObserver3.assertComplete();
  }

  @Test
  public void insertHashTest() {
    TestObserver<Void> testObserver = hashDao.insert(HASH_ENTITIES).test();

    testObserver.awaitTerminalEvent();
    testObserver.assertComplete();
  }

  @Test
  public void getHashTest() {
    TestObserver<Void> testObserver = hashDao.insert(HASH_ENTITIES).test();

    testObserver.awaitTerminalEvent();
    testObserver.assertComplete();

    TestObserver<List<HashEntity>> testObserver2 =
        hashDao
            .getDictionaryByLaoId(LAO_ID)
            .test()
            .assertValue(
                transactionObjects ->
                    transactionObjects.size() == 1
                        && transactionObjects.get(0).equals(HASH_ENTITY));

    testObserver2.awaitTerminalEvent();
    testObserver2.assertComplete();
  }

  @Test
  public void deleteHashByLaoTest() {
    TestObserver<Void> testObserver = hashDao.insert(HASH_ENTITIES).test();

    testObserver.awaitTerminalEvent();
    testObserver.assertComplete();

    TestObserver<Void> testObserver2 = hashDao.deleteByLaoId(LAO_ID).test();

    testObserver2.awaitTerminalEvent();
    testObserver2.assertComplete();

    // Assert that the transaction is actually deleted
    TestObserver<List<HashEntity>> testObserver3 =
        hashDao.getDictionaryByLaoId(LAO_ID).test().assertValue(List::isEmpty);

    testObserver3.awaitTerminalEvent();
    testObserver3.assertComplete();
  }
}

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/database/DigitalCashDatabaseTest.kt
package com.github.dedis.popstellar.repository.database

import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.github.dedis.popstellar.di.AppDatabaseModuleHelper.getAppDatabase
import com.github.dedis.popstellar.model.objects.Lao.Companion.generateLaoId
import com.github.dedis.popstellar.model.objects.digitalcash.TransactionObject
import com.github.dedis.popstellar.repository.DigitalCashRepositoryTest.Companion.getValidTransactionBuilder
import com.github.dedis.popstellar.repository.database.digitalcash.HashDao
import com.github.dedis.popstellar.repository.database.digitalcash.HashEntity
import com.github.dedis.popstellar.repository.database.digitalcash.TransactionDao
import com.github.dedis.popstellar.repository.database.digitalcash.TransactionEntity
import com.github.dedis.popstellar.testutils.Base64DataUtils
import io.reactivex.observers.TestObserver
import java.security.GeneralSecurityException
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class DigitalCashDatabaseTest {
  private lateinit var appDatabase: AppDatabase
  private lateinit var transactionDao: TransactionDao
  private lateinit var hashDao: HashDao

  private lateinit var TRANSACTION_OBJECT: TransactionObject
  private lateinit var TRANSACTION_ENTITY: TransactionEntity

  @Before
  @Throws(GeneralSecurityException::class)
  fun before() {
    appDatabase = getAppDatabase(ApplicationProvider.getApplicationContext())
    transactionDao = appDatabase.transactionDao()
    hashDao = appDatabase.hashDao()

    TRANSACTION_OBJECT = getValidTransactionBuilder("random id", USER1).build()
    TRANSACTION_ENTITY = TransactionEntity(LAO_ID, TRANSACTION_OBJECT)
  }

  @After
  fun close() {
    appDatabase.close()
  }

  @Test
  fun insertTransactionTest() {
    val testObserver = transactionDao.insert(TRANSACTION_ENTITY).test()
    testObserver.awaitTerminalEvent()
    testObserver.assertComplete()
  }

  @Test
  fun transactionsTest() {
    val testObserver = transactionDao.insert(TRANSACTION_ENTITY).test()
    testObserver.awaitTerminalEvent()
    testObserver.assertComplete()

    val testObserver2: TestObserver<List<TransactionObject>?> =
      transactionDao.getTransactionsByLaoId(LAO_ID).test().assertValue {
        transactionObjects: List<TransactionObject> ->
        (transactionObjects.size == 1 && transactionObjects[0] == TRANSACTION_OBJECT)
      }
    testObserver2.awaitTerminalEvent()
    testObserver2.assertComplete()
  }

  @Test
  fun deleteTransactionsByLaoTest() {
    val testObserver = transactionDao.insert(TRANSACTION_ENTITY).test()
    testObserver.awaitTerminalEvent()
    testObserver.assertComplete()

    val testObserver2 = transactionDao.deleteByLaoId(LAO_ID).test()
    testObserver2.awaitTerminalEvent()
    testObserver2.assertComplete()

    // Assert that the transaction is actually deleted
    val testObserver3: TestObserver<List<TransactionObject>?> =
      transactionDao.getTransactionsByLaoId(LAO_ID).test().assertValue {
        obj: List<TransactionObject?> ->
        obj.isEmpty()
      }
    testObserver3.awaitTerminalEvent()
    testObserver3.assertComplete()
  }

  @Test
  fun insertHashTest() {
    val testObserver = hashDao.insertAll(HASH_ENTITIES).test()
    testObserver.awaitTerminalEvent()
    testObserver.assertComplete()
  }

  @Test
  fun hashTest() {
    val testObserver = hashDao.insertAll(HASH_ENTITIES).test()
    testObserver.awaitTerminalEvent()
    testObserver.assertComplete()

    val testObserver2: TestObserver<List<HashEntity>?> =
      hashDao.getDictionaryByLaoId(LAO_ID).test().assertValue { transactionObjects: List<HashEntity>
        ->
        (transactionObjects.size == 1 && transactionObjects[0] == HASH_ENTITY)
      }
    testObserver2.awaitTerminalEvent()
    testObserver2.assertComplete()
  }

  @Test
  fun deleteHashByLaoTest() {
    val testObserver = hashDao.insertAll(HASH_ENTITIES).test()
    testObserver.awaitTerminalEvent()
    testObserver.assertComplete()

    val testObserver2 = hashDao.deleteByLaoId(LAO_ID).test()
    testObserver2.awaitTerminalEvent()
    testObserver2.assertComplete()

    // Assert that the transaction is actually deleted
    val testObserver3: TestObserver<List<HashEntity>?> =
      hashDao.getDictionaryByLaoId(LAO_ID).test().assertValue { obj: List<HashEntity?> ->
        obj.isEmpty()
      }
    testObserver3.awaitTerminalEvent()
    testObserver3.assertComplete()
  }

  companion object {
    private val LAO_ID = generateLaoId(Base64DataUtils.generatePublicKey(), 1000, "LAO1")
    private val LAO_ID2 = generateLaoId(Base64DataUtils.generatePublicKey(), 1500, "LAO2")
    private val USER1 = Base64DataUtils.generateKeyPair()
    private val HASH_ENTITY = HashEntity(USER1.publicKey.computeHash(), LAO_ID, USER1.publicKey)
    private val HASH_ENTITIES: List<HashEntity> = listOf(HASH_ENTITY)
  }
}
